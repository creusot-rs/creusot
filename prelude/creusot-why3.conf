[prover_modifiers]
name = "Z3"
driver = "z3_nombqi_seq.drv"