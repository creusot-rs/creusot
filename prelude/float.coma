module Float32
  use int.Int
  use bv.Pow2int
  use real.Abs as Abs
  use real.FromInt as FromInt
  use real.Truncate as Truncate
  use real.RealInfix
  use export ieee_float.RoundingMode

  type t = < float 8 24 >

  (** {2 Part I - Public Interface}   *)

  constant eb : int = t'eb
  (** the number of bits in the exponent.  *)

  constant sb : int = t'sb
  (** the number of bits in the significand, including the hidden bit. *)

  axiom eb_gt_1: 1 < eb
  axiom sb_gt_1: 1 < sb

  (** {3 Constructors and Constants} *)

  constant zeroF     : t   (** +0.0 *)
  (* exp_bias = 2^(eb - 1) - 1 *)
  (* max_finite_exp = 2^sb - 2 - exp_bias = exp_bias *)
  (* max_significand = (2^eb + 2^eb - 1) * 2^(1-eb) *)
  (* max_value = (2^eb + 2^eb - 1) * 2^(1-eb) * 2 ^ max_finite_exp *)
  (* [m;x] = ( 1 + m * 2^(1-eb) ) * 2^( x - exp_bias ) *)
  (* max_value = [2^(eb-1); 2^sb - 2] *)

  (** {3 Operators} *)
  let addi (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function addi mode t t : t
  let subi (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function subi mode t t : t
  let muli (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function muli mode t t : t
  let divi (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function divi mode t t : t
    (** The four basic operations, rounded in the given mode *)


  let abs (a : t) (ret (result : t)) = any
  function abs         t   : t   (** Absolute value *)

  let fma  (mode : mode) (a : t) (b : t) (c : t) (ret (result : t)) = any
  function fma  mode t t t : t   (** Fused multiply-add: x * y + z *)
  let sqrt (mode : mode) (a : t) (ret (result : t)) = any
  function sqrt mode   t   : t   (** Square root *)

  function neg (x : t) : t
  let neg (x : t) (ret (result : t)) = ret {neg x}

  function add (x : t) (y : t) : t = addi RNE x y
  let add (a : t) (b : t) (ret (result : t)) = ret {add a b}

  function sub (x : t) (y : t) : t = subi RNE x y
  let sub (a : t) (b : t) (ret (result : t)) = ret {sub a b}

  function mul (x : t) (y : t) : t = muli RNE x y
  let mul (a : t) (b : t) (ret (result : t)) = ret {mul a b}

  function div (x : t) (y : t) : t = divi RNE x y
  let div (a : t) (b : t) (ret (result : t)) = ret {div a b}

    (** Notations for operations in the default mode RNE *)

  function roundToIntegral mode t : t
  let roundToIntegral (mode : mode) (a : t) (ret (result : t) {result = roundToIntegral mode a}) = any
    (** Rounding to an integer *)

  function min       t t : t
  function max       t t : t

  predicate le t t
  predicate lt t t
  predicate eq t t
  predicate ge (x:t) (y:t) = le y x
  predicate gt (x:t) (y:t) = lt y x

  let le (a : t) (b : t) (ret (result : bool) { result = le a b }) = any
  let lt (a : t) (b : t) (ret (result : bool) { result = lt a b }) = any
  let eq (a : t) (b : t) (ret (result : bool) { result = eq a b }) = any
  let ge (a : t) (b : t) (ret (result : bool) { result = ge a b }) = any
  let gt (a : t) (b : t) (ret (result : bool) { result = gt a b }) = any
end
module Float64
  use int.Int
  use bv.Pow2int
  use real.Abs as Abs
  use real.FromInt as FromInt
  use real.Truncate as Truncate
  use real.RealInfix
  use export ieee_float.RoundingMode

  type t = < float 11 53 >

  (** {2 Part I - Public Interface}   *)

  constant eb : int = t'eb
  (** the number of bits in the exponent.  *)

  constant sb : int = t'sb
  (** the number of bits in the significand, including the hidden bit. *)

  axiom eb_gt_1: 1 < eb
  axiom sb_gt_1: 1 < sb

  (** {3 Constructors and Constants} *)

  constant zeroF     : t   (** +0.0 *)
  (* exp_bias = 2^(eb - 1) - 1 *)
  (* max_finite_exp = 2^sb - 2 - exp_bias = exp_bias *)
  (* max_significand = (2^eb + 2^eb - 1) * 2^(1-eb) *)
  (* max_value = (2^eb + 2^eb - 1) * 2^(1-eb) * 2 ^ max_finite_exp *)
  (* [m;x] = ( 1 + m * 2^(1-eb) ) * 2^( x - exp_bias ) *)
  (* max_value = [2^(eb-1); 2^sb - 2] *)

  (** {3 Operators} *)
  let addi (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function addi mode t t : t
  let subi (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function subi mode t t : t
  let muli (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function muli mode t t : t
  let divi (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function divi mode t t : t
    (** The four basic operations, rounded in the given mode *)


  let abs (a : t) (ret (result : t)) = any
  function abs         t   : t   (** Absolute value *)

  let fma  (mode : mode) (a : t) (b : t) (c : t) (ret (result : t)) = any
  function fma  mode t t t : t   (** Fused multiply-add: x * y + z *)
  let sqrt (mode : mode) (a : t) (ret (result : t)) = any
  function sqrt mode   t   : t   (** Square root *)

  function neg (x : t) : t
  let neg (x : t) (ret (result : t)) = ret {neg x}

  function add (x : t) (y : t) : t = addi RNE x y
  let add (a : t) (b : t) (ret (result : t)) = ret {add a b}

  function sub (x : t) (y : t) : t = subi RNE x y
  let sub (a : t) (b : t) (ret (result : t)) = ret {sub a b}

  function mul (x : t) (y : t) : t = muli RNE x y
  let mul (a : t) (b : t) (ret (result : t)) = ret {mul a b}

  function div (x : t) (y : t) : t = divi RNE x y
  let div (a : t) (b : t) (ret (result : t)) = ret {div a b}

    (** Notations for operations in the default mode RNE *)

  function roundToIntegral mode t : t
  let roundToIntegral (mode : mode) (a : t) (ret (result : t) {result = roundToIntegral mode a}) = any
    (** Rounding to an integer *)

  function min       t t : t
  function max       t t : t

  predicate le t t
  predicate lt t t
  predicate eq t t
  predicate ge (x:t) (y:t) = le y x
  predicate gt (x:t) (y:t) = lt y x

  let le (a : t) (b : t) (ret (result : bool) { result = le a b }) = any
  let lt (a : t) (b : t) (ret (result : bool) { result = lt a b }) = any
  let eq (a : t) (b : t) (ret (result : bool) { result = eq a b }) = any
  let ge (a : t) (b : t) (ret (result : bool) { result = ge a b }) = any
  let gt (a : t) (b : t) (ret (result : bool) { result = gt a b }) = any
end