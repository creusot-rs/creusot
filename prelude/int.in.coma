module UInt$bits_count$
  use mach.int.Int

  type t = < range 0 $max_unsigned_value$ >

  constant two_power_size : int = $two_power_size$

  axiom extensionality:
    forall x y: t [t'int x, t'int y]. t'int x = t'int y -> x = y

  predicate in_bounds [@inline:trivial] (n:int) = 0 <= n <= $max_unsigned_value$

  function of_int (n: int) : t
  axiom to_of_int:
    forall x: int [of_int x]. in_bounds x -> t'int (of_int x) = x

  function add (a:t) (b:t) : t
  axiom add_in_bounds:
    forall x y:t [add x y]. in_bounds(t'int x + t'int y) -> t'int(add x y) = t'int x + t'int y
  axiom add_wrapping:
    forall x y:t [add x y]. exists k:int. t'int(add x y) = t'int x + t'int y + k * two_power_size

  function add_with_overflow (a:t) (b:t) : (t, bool)
  axiom add_with_overflow_spec : forall x y:t. add_with_overflow x y = (add x y, false = in_bounds(t'int x + t'int y))

  function sub (a:t) (b:t) : t
  axiom sub_in_bounds:
    forall x y:t [sub x y]. in_bounds(t'int x - t'int y) -> t'int(sub x y) = t'int x - t'int y
  axiom sub_wrapping:
    forall x y:t [sub x y]. exists k:int. t'int(sub x y) = t'int x - t'int y + k * two_power_size

  function mul (a:t) (b:t) : t
  axiom mul_in_bounds:
    forall x y:t [mul x y]. in_bounds(t'int x * t'int y) -> t'int(mul x y) = t'int x * t'int y
  axiom mul_wrapping:
    forall x y:t [mul x y]. exists k:int. t'int(mul x y) = t'int x * t'int y + k * two_power_size

  function neg (a:t): t
  axiom neg_zero:
    forall x:t [neg x]. t'int x = 0 -> t'int(neg x) = 0
  axiom neg_not_zero:
    forall x:t [neg x]. 0 < t'int x <= $max_unsigned_value$ -> t'int(neg x) = two_power_size - t'int x

  function of_bool [@inline:trivial] (b:bool) : t = if b then 1:t else 0:t

  predicate le [@inline:trivial] (a:t) (b:t) = t'int a <= t'int b
  predicate lt [@inline:trivial] (a:t) (b:t) = t'int a < t'int b
  predicate ge [@inline:trivial] (a:t) (b:t) = t'int a >= t'int b
  predicate gt [@inline:trivial] (a:t) (b:t) = t'int a > t'int b

  function bw_and (a:t) (b:t) : t
  function bw_or (a:t) (b:t) : t
  function bw_xor (a:t) (b:t) : t
  function bw_not (a:t) : t

  let of_int (n:int) { [@expl:integer overflow] in_bounds n }
    (ret (result : t) { t'int result = n }) = any

  let add (a:t) (b:t)
    { [@expl:integer overflow] in_bounds (t'int a + t'int b) }
    (ret (result : t) { t'int result = t'int a + t'int b })
    = any
  let sub (a:t) (b:t)
    { [@expl:integer overflow] in_bounds (t'int a - t'int b) }
    (ret (result : t) { t'int result = t'int a - t'int b })
    = any
  let mul (a:t) (b:t)
    { [@expl:integer overflow] in_bounds (t'int a * t'int b) }
    (ret (result : t) { t'int result = t'int a * t'int b })
    = any
  let div (a:t) (b:t)
    { [@expl:division by zero] t'int b <> 0 }
    (ret (result : t) { t'int result = div (t'int a) (t'int b) })
    = any
  let rem (a:t) (b:t)
    { [@expl:division by zero] t'int b <> 0 }
    (ret (result : t) { t'int result = mod (t'int a) (t'int b) })
    = any
  let shl (a:t) (b:int)
    { [@expl:out-of-bounds shifting] b >= 0 /\ b <= $bits_count$  }
    (ret (result :t) { true })
    = any
  let shr (a:t) (b:int)
    { [@expl:out-of-bounds shifting] b >= 0 /\ b <= $bits_count$ }
    (ret (result :t) { true })
    = any
end

module UInt$bits_count$BW
  use export bv.BV$bits_count$
  use bv.BV256 as BV256
  use bv.BVConverter_$bits_count$_256
  use mach.int.Int

  constant max_uint_as_BV256 : BV256.t = $max_unsigned_value$

  function to_BV256 (x: t) : BV256.t = toBig x
  function of_BV256 (x: BV256.t) : t = toSmall x
  function of_bool [@inline:trivial] (b : bool) : t = if b then 1:t else 0:t

  predicate le (a:t) (b:t) = ule a b
  predicate lt (a:t) (b:t) = ult a b
  predicate ge (a:t) (b:t) = uge a b
  predicate gt (a:t) (b:t) = ugt a b

  let of_BV256 (a:BV256.t)
    { [@expl:arithmetic overflow] BV256.t'int a >= 0 /\ BV256.t'int a < two_power_size }
    (ret (result: t) { result = of_BV256 a })
    = any

  let add (a:t) (b:t)
    { [@expl:arithmetic overflow] t'int a + t'int b < two_power_size \/ BV256.ule (BV256.add (to_BV256 a) (to_BV256 b)) max_uint_as_BV256 }
    (ret (result :t)  { t'int result = t'int a + t'int b } { result = add a b })
    = any
  let sub (a:t)  (b:t)
    { [@expl:arithmetic overflow] t'int a >= t'int b \/ uge a b }
    (ret (result: t) { t'int result = t'int a - t'int b } { result = sub a b })
    = any
  let mul (a:t) (b:t)
    { [@expl:arithmetic overflow] t'int a * t'int b < two_power_size \/ BV256.ule (BV256.mul (to_BV256 a) (to_BV256 b)) max_uint_as_BV256 }
    (ret (result: t) { result = mul a b } { t'int result = t'int a * t'int b })
    = any
  let div (a:t) (b:t)
    { [@expl:division by zero] b <> zeros \/ t'int b <> 0 }
    (ret (result: t) { t'int result = div (t'int a) (t'int b) } { result = udiv a b })
    = any
  let rem (a:t) (b:t)
    { [@expl:remainder by zero] b <> zeros  \/ t'int b <> 0 }
    (ret (result: t) { t'int result = mod (t'int a) (t'int b) } { result = urem a b })
    = any
  let shl (a:t) (b:int)
    { [@expl:out-of-bounds shifting] b >= 0 /\ b <= size  }
    (ret (result :t) { result = lsl_bv a (of_int b) } { result = lsl a b })
    = any
  let shr (a:t) (b:int)
    { [@expl:out-of-bounds shifting] b >= 0 /\ b <= size }
    (ret (result :t) { result = lsr_bv a (of_int b) } { result = lsr a b })
    = any
end

module Int$bits_count$
  use mach.int.Int

  type t = < range -$min_signed_value$ $max_signed_value$ >

  constant min_int : int = - $min_signed_value$
  constant max_int : int =   $max_signed_value$
  constant two_power_size : int = $two_power_size$

  predicate in_bounds [@inline:trivial] (n:int) = min_int <= n <= max_int

  function to_int (x:t) : int = t'int x
  axiom extensionality:
    forall x y: t [to_int x, to_int y]. to_int x = to_int y -> x = y

  function of_int (n: int) : t
  axiom to_of_int:
    forall x: int [of_int x]. in_bounds x -> to_int (of_int x) = x

  function add (a:t) (b:t) : t
  axiom add_in_bounds:
    forall x y:t [add x y]. in_bounds(to_int x + to_int y) -> to_int(add x y) = to_int x + to_int y
  axiom add_wrapping:
    forall x y:t [add x y]. exists k:int. to_int(add x y) = to_int x + to_int y + k * two_power_size

  function sub (a:t) (b:t) : t
  axiom sub_in_bounds:
    forall x y:t [sub x y]. in_bounds(to_int x - to_int y) -> to_int(sub x y) = to_int x - to_int y
  axiom sub_wrapping:
    forall x y:t [sub x y]. exists k:int. to_int(sub x y) = to_int x - to_int y + k * two_power_size

  function mul (a:t) (b:t) : t
  axiom mul_in_bounds:
    forall x y:t [mul x y]. in_bounds(to_int x * to_int y) -> to_int(mul x y) = to_int x * to_int y
  axiom mul_wrapping:
    forall x y:t [mul x y]. exists k:int. to_int(mul x y) = to_int x * to_int y + k * two_power_size

  function neg (a:t): t
  axiom neg_in_bounds:
    forall x:t [neg x]. min_int < to_int x <= max_int -> to_int(neg x) = - (to_int x)
  axiom neg_wrapping:
    forall x:t [neg x]. to_int x = min_int -> to_int(neg x) = min_int

  function of_bool [@inline:trivial] (b:bool) : t = if b then 1:t else 0:t

  predicate le [@inline:trivial] (a:t) (b:t) = to_int a <= to_int b
  predicate lt [@inline:trivial] (a:t) (b:t) = to_int a < to_int b
  predicate ge [@inline:trivial] (a:t) (b:t) = to_int a >= to_int b
  predicate gt [@inline:trivial] (a:t) (b:t) = to_int a > to_int b

  function bw_and (a:t) (b:t) : t
  function bw_or (a:t) (b:t) : t
  function bw_xor (a:t) (b:t) : t
  function bw_not (a:t) : t

  let of_int (n:int) { [@expl:integer overflow] in_bounds n }
    (ret (result : t) { to_int result = n }) = any
  let neg (a:t)
    { [@expl:integer overflow] to_int a <> min_int }
    (ret (result : t) { to_int result =  - (to_int a) })
    = any
  let add (a:t) (b:t)
    { [@expl:integer overflow] in_bounds (to_int a + to_int b) }
    (ret (result : t) { to_int result = to_int a + to_int b })
    = any
  let sub (a:t) (b:t)
    { [@expl:integer overflow] in_bounds (to_int a - to_int b) }
    (ret (result : t) { to_int result = to_int a - to_int b })
    = any
  let mul (a:t) (b:t)
    { [@expl:integer overflow] in_bounds (to_int a * to_int b) }
    (ret (result : t) { to_int result = to_int a * to_int b })
    = any
  let div (a:t) (b:t)
    { [@expl:division by zero] (to_int b) <> 0 }
    { [@expl:integer overflow] not (to_int a = min_int /\ to_int b = -1) }
    (ret (result : t) { to_int result = div (to_int a) (to_int b) })
    = any
  let rem (a:t) (b:t)
    { [@expl:division by zero] (to_int b) <> 0 }
    { [@expl:integer overflow] not (to_int a = min_int /\ to_int b = -1) }
    (ret (result : t) { true })
    = any
  let shl (a:t) (b:int)
    { [@expl:out-of-bounds shifting] b >= 0 /\ b <= $bits_count$  }
    (ret (result :t) { true })
    = any
  let shr (a:t) (b:int)
    { [@expl:out-of-bounds shifting] b >= 0 /\ b <= $bits_count$ }
    (ret (result :t) { true })
    = any
end

module Int$bits_count$BW
  use export bv.BV$bits_count$
  use bv.BV256 as BV256
  use bv.BVConverter_$bits_count$_256
  use mach.int.Int

  constant min_sint : t = $min_signed_value$
  constant minus_one : t = $max_unsigned_value$
  constant min_sint_as_BV256 : BV256.t = $min_signed_value$
  constant max_sint_as_BV256 : BV256.t = $max_signed_value$

  function to_BV256 (x: t) : BV256.t = stoBig x
  function of_BV256 (x: BV256.t) : t = toSmall x

  function of_bool [@inline:trivial] (b : bool) : t = if b then 1:t else 0:t

  predicate le (a:t) (b:t) = sle a b
  predicate lt (a:t) (b:t) = slt a b
  predicate ge (a:t) (b:t) = sge a b
  predicate gt (a:t) (b:t) = sgt a b

  let of_BV256 (a:BV256.t)
    { [@expl:arithmetic overflow] BV256.to_int a >= 0 /\ BV256.to_int a < two_power_size }
    (ret (result: t) { result = of_BV256 a })
    = any
  let neg (a:t)
    { [@expl:arithmetic overflow] a <> min_sint }
    (ret (result :t)  { to_int result = - to_int a  } { result = neg a})
    = any
  let add (a:t) (b:t)
    { [@expl:arithmetic overflow] - two_power_size_minus_one <= to_int a + to_int b < two_power_size_minus_one \/ let r = BV256.add (to_BV256 a) (to_BV256 b) in (BV256.sle min_sint_as_BV256 r /\ BV256.sle r max_sint_as_BV256) }
    (ret (result :t)  { to_int result = to_int a + to_int b } { result = add a b })
    = any
  let sub (a:t)  (b:t)
    { [@expl:arithmetic overflow] - two_power_size_minus_one <= to_int a - to_int b < two_power_size_minus_one \/ let r = BV256.sub (to_BV256 a) (to_BV256 b) in (BV256.sle min_sint_as_BV256 r /\ BV256.sle r max_sint_as_BV256) }
    (ret (result: t) { to_int result = to_int a - to_int b } { result = sub a b })
    = any
  let mul (a:t) (b:t)
    { [@expl:arithmetic overflow] - two_power_size_minus_one <= to_int a * to_int b < two_power_size_minus_one \/ let r = BV256.mul (to_BV256 a) (to_BV256 b) in (BV256.sle min_sint_as_BV256 r /\ BV256.sle r max_sint_as_BV256) }
    (ret (result: t) { to_int result = to_int a * to_int b } { result = mul a b })
    = any
  let div (a:t) (b:t)
    { [@expl:division by zero] b <> zeros \/ to_int b <> 0 }
    { [@expl:signed division overflow check] (a <> min_sint \/ b <> minus_one) \/ (to_int a <> to_int min_sint \/ to_int b <> -1) }
    (ret (result: t) { to_int result = div (to_int a) (to_int b) } { result = sdiv a b })
    = any
  let rem (a:t) (b:t)
    { [@expl:remainder by zero] b <> zeros \/ to_int b <> 0 }
    (ret (result: t) { to_int result = mod (to_int a) (to_int b) } { result = srem a b })
    = any
  let shl (a:t) (b:int)
    { [@expl:out-of-bounds shifting] b >= 0 /\ b < size  }
    (ret (result :t) { result = lsl_bv a (of_int b) } { result = lsl a b })
    = any
  let shr (a:t) (b:int)
    { [@expl:out-of-bounds shifting] b >= 0 /\ b < size }
    (ret (result :t) { result = asr_bv a (of_int b) } { result = asr a b })
    = any
end
