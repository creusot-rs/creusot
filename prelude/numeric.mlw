module BoundedInt

  use int.Int

  type t

  constant min : int
  constant max : int

  function to_int (n:t) : int
  meta coercion function to_int
  meta "model_projection" function to_int

  val to_int (n:t) : int
    ensures { result = n }

  predicate in_bounds (n:int) = min <= n <= max

  axiom to_int_in_bounds: forall n:t. in_bounds n

  val of_int (n:int) : t
    requires { [@expl:integer overflow] in_bounds n }
    ensures  { result = n }

  val function add (a:t) (b:t) : t
    requires { [@expl:integer overflow] in_bounds (a + b) }
    ensures   { result = a + b }

  val function sub (a:t) (b:t) : t
    requires { [@expl:integer overflow] in_bounds (a - b) }
    ensures  { result = a - b }

  val function mul (a:t) (b:t) : t
    requires { [@expl:integer overflow] in_bounds (a * b) }
    ensures  { result = a * b }

  val function neg (a:t) : t
    requires { [@expl:integer overflow] in_bounds (- a) }
    ensures  { result = - a }

  axiom extensionality: forall x y: t. to_int x = to_int y -> x = y

  val function eq (a:t) (b:t) : bool
    ensures { result <-> a = b }
    ensures { to_int a = to_int b -> result }

  val function ne (a:t) (b:t) : bool
    ensures { result <-> a <> b }
    ensures { to_int a <> to_int b -> result }

  val function le (a:t) (b:t) : bool
    ensures  { result <-> to_int a <= to_int b }

  val function lt (a:t) (b:t) : bool
    ensures  { result <-> to_int a < to_int b }

  val function ge (a:t) (b:t) : bool
    ensures  { result <-> to_int a >= to_int b }

  val function gt (a:t) (b:t) : bool
    ensures  { result <-> to_int a > to_int b }

  use int.ComputerDivision

  val function div (a:t) (b:t) : t
    requires { [@expl:division by zero] b <> 0 }
    requires { [@expl:integer overflow] in_bounds (div a b) }
    ensures  { result = div a b }

  val function rem (a:t) (b:t) : t
    requires { [@expl:division by zero] b <> 0 }
    requires { [@expl:integer overflow] in_bounds (mod a b) }
    ensures  { result = mod a b }

end

module Unsigned

  use int.Int

  let constant min_unsigned : int = 0

  clone export BoundedInt with
    constant min = min_unsigned, axiom .

  constant zero_unsigned : t

  axiom zero_unsigned_is_zero : to_int zero_unsigned = 0

  constant radix : int

  axiom radix_def : radix = max+1

end