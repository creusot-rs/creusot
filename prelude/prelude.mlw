module Prelude
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double

  type usize = uint64
  type isize = int64
  type opaque_ptr

  type borrowed 'a = { current : 'a ; final : 'a; }
  let function ( *_ ) x = x.current
  let function ( ^_ ) x = x.final
  val borrow_mut (a : 'a) : borrowed 'a
  ensures { *result = a }
end
