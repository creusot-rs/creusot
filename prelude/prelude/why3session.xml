<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE why3session PUBLIC "-//Why3//proof session v5//EN"
"http://why3.lri.fr/why3session.dtd">
<why3session shape_version="6">
<file format="whyml" proved="true">
<path name=".."/><path name="prelude.mlw"/>
</file>
</why3session>
