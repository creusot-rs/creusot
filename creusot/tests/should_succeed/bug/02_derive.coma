
module C02Derive_Lit_Type
  type t_lit  =
    | C_Lit
    
  function any_l (_ : 'b) : 'a
  let rec t_lit (input:t_lit) (ret  )= any [ good -> {C_Lit  = input} (! ret) | bad -> {C_Lit  <> input} {false} any ] 
end
module C02Derive_Impl0_Clone
  use prelude.Borrow
  use prelude.Intrinsic
  use C02Derive_Lit_Type as C02Derive_Lit_Type
  let rec clone' (self:C02Derive_Lit_Type.t_lit) (return'  (ret:C02Derive_Lit_Type.t_lit))= (! bb0
    [ bb0 = s0 [ s0 =  [ &_0 <- C02Derive_Lit_Type.C_Lit ] s1 | s1 = return' {_0} ]  ]
    ) [ & _0 : C02Derive_Lit_Type.t_lit = any_l () : C02Derive_Lit_Type.t_lit ] 
    [ return' (result:C02Derive_Lit_Type.t_lit)-> (! return' {result}) ]
    
end
module C02Derive_Impl0
  use C02Derive_Lit_Type as C02Derive_Lit_Type
  predicate invariant'1 (self : C02Derive_Lit_Type.t_lit) =
    [#"../../../../../creusot-contracts/src/invariant.rs" 8 8 8 12] true
  let rec invariant'1 (self:C02Derive_Lit_Type.t_lit) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = invariant'1 self} (! return' {result}) ]
    
  predicate inv'1 (_x : C02Derive_Lit_Type.t_lit)
  let rec inv'1 (_x:C02Derive_Lit_Type.t_lit) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = inv'1 _x} (! return' {result}) ]
    
  axiom inv'1 : forall x : C02Derive_Lit_Type.t_lit . inv'1 x = true
  predicate invariant'0 (self : C02Derive_Lit_Type.t_lit) =
    [#"../../../../../creusot-contracts/src/invariant.rs" 8 8 8 12] true
  let rec invariant'0 (self:C02Derive_Lit_Type.t_lit) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = invariant'0 self} (! return' {result}) ]
    
  predicate inv'0 (_x : C02Derive_Lit_Type.t_lit)
  let rec inv'0 (_x:C02Derive_Lit_Type.t_lit) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = inv'0 _x} (! return' {result}) ]
    
  axiom inv'0 : forall x : C02Derive_Lit_Type.t_lit . inv'0 x = true
  use prelude.Borrow
  goal clone'_refn : [#"../02_derive.rs" 3 9 3 14] forall self : C02Derive_Lit_Type.t_lit . inv'0 self
   -> (forall result : C02Derive_Lit_Type.t_lit . inv'1 result /\ result = self)
end
