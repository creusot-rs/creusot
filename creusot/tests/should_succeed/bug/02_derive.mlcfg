
module C02Derive_Lit_Type
  type t_lit  =
    | C_Lit
    
end
module C02Derive_Impl0_Clone_Interface
  use prelude.Borrow
  use C02Derive_Lit_Type as C02Derive_Lit_Type
  val clone' [#"../02_derive.rs" 3 9 3 14] (self : C02Derive_Lit_Type.t_lit) : C02Derive_Lit_Type.t_lit
end
module C02Derive_Impl0_Clone
  use prelude.Borrow
  use C02Derive_Lit_Type as C02Derive_Lit_Type
  let rec cfg clone' [#"../02_derive.rs" 3 9 3 14] [@cfg:stackify] [@cfg:subregion_analysis] (self : C02Derive_Lit_Type.t_lit) : C02Derive_Lit_Type.t_lit
    
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : C02Derive_Lit_Type.t_lit;
  {
    goto BB0
  }
  BB0 {
    _0 <- C02Derive_Lit_Type.C_Lit;
    return _0
  }
  
end
module CreusotContracts_Invariant_Inv_Stub
  type t
  predicate inv (_x : t)
end
module CreusotContracts_Invariant_Inv_Interface
  type t
  predicate inv (_x : t)
  val inv (_x : t) : bool
    ensures { result = inv _x }
    
end
module CreusotContracts_Invariant_Inv
  type t
  predicate inv (_x : t) =
    [#"../../../../../creusot-contracts/src/invariant.rs" 27 4 27 8] true
  val inv (_x : t) : bool
    ensures { result = inv _x }
    
end
module TyInv_Trivial
  type t
  clone CreusotContracts_Invariant_Inv_Stub as Inv0 with
    type t = t
  axiom inv_trivial : forall self : t . Inv0.inv self = true
end
module C02Derive_Impl0
  use prelude.Borrow
  use C02Derive_Lit_Type as C02Derive_Lit_Type
  clone CreusotContracts_Invariant_Inv_Interface as Inv1 with
    type t = C02Derive_Lit_Type.t_lit
  clone TyInv_Trivial as TyInv_Trivial1 with
    type t = C02Derive_Lit_Type.t_lit,
    predicate Inv0.inv = Inv1.inv,
    axiom .
  clone CreusotContracts_Invariant_Inv_Interface as Inv0 with
    type t = C02Derive_Lit_Type.t_lit
  clone TyInv_Trivial as TyInv_Trivial0 with
    type t = C02Derive_Lit_Type.t_lit,
    predicate Inv0.inv = Inv0.inv,
    axiom .
  goal clone'_refn : [#"../02_derive.rs" 3 9 3 14] forall self : C02Derive_Lit_Type.t_lit . Inv0.inv self -> (forall result : C02Derive_Lit_Type.t_lit . Inv1.inv result /\ result = self)
end
