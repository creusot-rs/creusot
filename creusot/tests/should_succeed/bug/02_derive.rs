extern crate creusot_contracts;

#[derive(Clone)]
pub struct Lit {}
