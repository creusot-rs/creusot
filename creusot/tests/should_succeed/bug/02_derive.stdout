module Type
  use Ref
  use mach.int.Int
  use prelude.Int8
  use prelude.Int16
  use mach.int.Int32
  use mach.int.Int64
  use prelude.UInt8
  use prelude.UInt16
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
  type c02derive_lit  = 
    | C02Derive_Lit
    
end
module Core_Clone_Clone_Clone_Interface
  type self   
  use prelude.Prelude
  val clone' [@cfg:stackify] (self : self) : self
    requires {false}
    
end
module Core_Clone_Clone_Clone
  type self   
  use prelude.Prelude
  val clone' [@cfg:stackify] (self : self) : self
    requires {false}
    
end
module Core_Clone_Clone_CloneFrom_Interface
  type self   
  use prelude.Prelude
  val clone_from [@cfg:stackify] (self : borrowed self) (source : self) : ()
    requires {false}
    
end
module Core_Clone_Clone_CloneFrom
  type self   
  use prelude.Prelude
  val clone_from [@cfg:stackify] (self : borrowed self) (source : self) : ()
    requires {false}
    
end
module C02Derive_Impl0_Clone_Interface
  use prelude.Prelude
  use Type
  val clone' [@cfg:stackify] (self : Type.c02derive_lit) : Type.c02derive_lit
end
module C02Derive_Impl0_Clone
  use prelude.Prelude
  use Type
  let rec cfg clone' [@cfg:stackify] (self : Type.c02derive_lit) : Type.c02derive_lit = 
  var _0 : Type.c02derive_lit;
  var self_1 : Type.c02derive_lit;
  {
    self_1 <- self;
    goto BB0
  }
  BB0 {
    assume { (fun x -> true) self_1 };
    _0 <- Type.C02Derive_Lit;
    return _0
  }
  
end
module C02Derive_Impl0
  use Type
  clone C02Derive_Impl0_Clone_Interface as Clone0
  clone Core_Clone_Clone_Clone_Interface as Clone1 with type self = Type.c02derive_lit, val clone' = Clone0.clone'
end
